controllers/toolchaincluster_resources/toolchaincluster_resources_controller.go

```diff
@@ -0,0 +1,77 @@
+package toolchaincluster_resources
+
+import (
+	"context"
+	"embed"
+	"fmt"
+
+	toolchainv1alpha1 "github.com/codeready-toolchain/api/api/v1alpha1"
+	commoncontroller "github.com/codeready-toolchain/toolchain-common/controllers"
+	applycl "github.com/codeready-toolchain/toolchain-common/pkg/client"
+	commonpredicates "github.com/codeready-toolchain/toolchain-common/pkg/predicate"
+	"github.com/codeready-toolchain/toolchain-common/pkg/template"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	runtimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// ResourceControllerLabelValue is being added to all the resources managed by this controller.
+// It's then used to filter all the events on those resources by using a mapper function in the watcher configuration.
+const ResourceControllerLabelValue = "toolchaincluster-resources-controller" // todo move this label value to api repo
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *Reconciler) SetupWithManager(mgr ctrl.Manager, operatorNamespace string) error {
+	// check for required templates FS directory
+	if r.templates == nil {
+		return fmt.Errorf("no templates FS configured")
+	}
+
+	build := ctrl.NewControllerManagedBy(mgr).
+		For(&v1.ServiceAccount{})
+
+	// add watcher for all kinds from given templates
+	var err error
+	r.templateObjects, err = template.LoadObjectsFromEmbedFS(r.templates, &template.Variables{Namespace: operatorNamespace})
+	if err != nil {
+		return err
+	}
+	mapToOwnerByLabel := handler.EnqueueRequestsFromMapFunc(commoncontroller.MapToControllerByMatchingLabel(operatorNamespace, toolchainv1alpha1.ProviderLabelKey, ResourceControllerLabelValue))
+	for _, obj := range r.templateObjects {
+		build = build.Watches(&source.Kind{Type: obj.DeepCopyObject().(runtimeclient.Object)}, mapToOwnerByLabel, builder.WithPredicates(commonpredicates.LabelsAndGenerationPredicate{}))
```

Comment:
Is there some way to set up the controller to only watch for resources from a single namespace? I think the total number of serviceaccounts+roles+clusterroles+rolebindings in the cluster can get quite big, so we should make our controller as cache-friendly as possible.

Reply:
This is configured in the manager , so for example for host operator it will watch and cache objects in the toolchain-host-operator namespace https://github.com/codeready-toolchain/host-operator/blob/de95481affea496ceb043113a347f64a61abe42c/main.go#L205. 

This is my understanding. So it should not watch and cache objects in all namespaces.

Reply:
Yes, by default we use a cache scoped to the operator's namespace only, and only when it's really necessary we use a client with a cache scoped for all namespaces - eg: https://github.com/codeready-toolchain/member-operator/blob/cacdbe4681bccddaf7c0d1c9957e97eb8fcfee92/main.go#L188